### DISCUSSION TITLE
Ambiguous user defined conversions and inheritance

### DISCUSSION THREAD

**Discussion by dzmitry-lahoda** [dotnet/csharplang#6050] [General]
# Issue

The compiler cannot infer proper cast in the next 2 cases:

```csharp
    class A : Exception{}
    class B:A{}
    record C()
    {
        public static implicit operator A(C self) => throw new Exception();
    }
    record D(): C
    {
        public static implicit operator B(D self) => throw new Exception();
    }

    record Main() {
        static Main() {
            var d = new D();
            // Ambiguous user defined conversions 'D.implicit operator B(D)' and 'C.implicit operator A(C)' when converting from 'D' to 'Exception' [Matchmaking]csharp(CS0457)
            throw (A)d;
        }
        // Ambiguous user defined conversions 'D.implicit operator B(D)' and 'C.implicit operator A(C)' when converting from 'D' to 'Exception' [Matchmaking]csharp(CS0457)
        static A a => new D();
        static B b => new D();
    }
```

# Question

Could the compiler compile `throw (A)d;`  and/or `static A a => new D();` because I directly asked it what typecast I want?

# Why

C# has no multiple inheritances, but it got interfaces with implementations, extensions, and static interfaces.  

But sometimes I feel the need to express multiple inheritances so. Sometimes I can use implicit casts. 

So also they feel like if I have inherited, for some reason compiler cannot decide what cast to choose even if it has direct suggestions of what to choose.  

If cast can be decided, it can make the lack of multiple inheritances even less visible.

# Notes

I have seen similar cases where put into `csharplang`, not `roslyn`. So I have put it there too.

**Comment by CyrusNajmabadi** [comment]
> because I directly asked it what typecast I want?

You directly asked, but there is no corresponding conversion operator defined for that.  Instead, there are two potential *different* conversions that could lead to that.  Specifically:

`(A)(B)d` or `(A)(C)d`

So it doesn't know which is better and gives up.

**Reply by dzmitry-lahoda** [reply]
I see. I would prefer `(A)(C)d`, so 

1. check if can use inheritance to cast
2. if no, then use the operator

The assumption is that inheritance is done first, and then people add casts. So adding casts will not break inheritance rule.

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda why not just add the casts in explicitly to make it clear what you want?  A core problem i had looking at your original code is that i myself had no idea what it should mean.  ANd that means the API design itself is very suspect as it would likely be hard to reason about and coudl easily lead to unexpected outcomes.

**Reply by dzmitry-lahoda** [reply]
So explicit cast operators will work and has no conflict?

I am trying to have 2 class hierarchies which is easy to switch from ine to other.

**Reply by dzmitry-lahoda** [reply]
Or explicitly means typing '(A)(C)d'? It is ugly.

**Reply by CyrusNajmabadi** [reply]
> Or explicitly means typing '(A)(C)d'? It is ugly.

I know exactly what the code means when you write that.  I'd rather have clear and easy to understand code vs code that is highly complex and difficult to understand :)

> I am trying to have 2 class hierarchies which is easy to switch from ine to other.

Consider an explicit method call to do that.  Haivn this happen implicitly is, as you can see, extremely confusing :)

**Reply by dzmitry-lahoda** [reply]
I am thinking about method calls, but still cannot make them nice and convenient, and at the same time very generic/typed and seamless.

Attempts:
1. https://github.com/dotnet/csharplang/issues/1628#issuecomment-1106621156
2. I have Exceptions in my code as second tree of inheritance (ASP.NET ProblemDetails result <-> ProblemDetailsException : Exception).
```csharp
     [DoesNotReturn]
     public virtual void Throw() => throw new ProblemDetailsException(this);
```


What I see in one case C# explicitly prefers inheritance overcast.
```csharp
        record A();
        record B(): A
        {
            // 'B.implicit operator A(B)': user-defined conversions to or from a base type are not allowed [main]csharp(CS0553)
            public static implicit operator A(B self) => throw new NotImplementedException();
        };
```

For me more complicated, but predictable rules I not highly complex or difficult to understand(actually there is Rust language that does that). For me reading code with typecast noise and reflection hacks or irrelevant properties is harder (to have multiple inheritance-like behaviors).

The rule 'prefer inheritance over operators' sounds sane.

**Reply by CyrusNajmabadi** [reply]
> The rule 'prefer inheritance over operators' sounds sane.

As i said, looking at the original code, i have no idea what it meant.  So it seems unlikely we would go down this path.

**Reply by dzmitry-lahoda** [reply]
this
```
static A a => new D();
```
is compiled as
```
static A a => (C)(new D());
```

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda that interpretation doesn't seem any more reasonable or desirable over `static A a => (B)new D();`.  Both involve a conversion and inheritance.  It's just if conversion happens first, then inheritance, or inheritance than conversion.  Neither seems particularly better, and it's extremely confusing given that both seem totally reasonable interpretations.  Given multiple interpretations that are all reasonable, having this be ambiguous seems like a good thing.  The code should be explicit about exactly what is going on here so that readers (like myself) understand it.

WRT to this discussion, something that is lacking is a compelling use case.  The example provied is just too abstract, and i have no intuition on what A, B, C or D represent or why this particular interpretation would be in any way the intuitive/desirable one.  For all i know there's another example that showed that the opposite interpretation would be far more desirable.  In that regard, it feels like attempting to relax a rule that helps prevent confusion just for your particular use case (which hasn't been shown).

**Reply by dzmitry-lahoda** [reply]
Imagine a path from A to F. There are several paths to reach from A to F. There could be several casts chosen on each step to the final type F. But because there is one and only one possible inheritance on each step the path decision on each step if the inheritance chosen is unique. 

Ambiguity appears if there is no inheritance, but several implicit to choose from. 

So rule can be
```
1. prefer inheritance over operators. only one inheritance is possible on each step.
2. if there are several implicit on one step, but no inheritance - fail compilation
```

So instead of having a set of implicit from the whole graph (current behavior), can really use graph traversal with the above rules.

Inheritance changes seldom - so introducing bad behavior is hard. 

And when new inheritance appears there is a much more subtle way to be unpredictable on reflection and get issues.

I would rather debug the compiler in compile time (operators and inheritance), than reflection.

Why reflection? Because if there are no multiple inheritances then people will use reflection a lot.  In my case inherit and hide members,  make different properties with the same baked data but with different serde attributes.

I tell that reflection(or copy-paste) is more complex(harder to debug, easier to get copy-paste out of date code issues) than inheritance rule.

So I do not think you showed desirable, but more of opinions, as I did too.

What you could do is prove that the inheritance rule is unsound and will not work. That kind of talk I expect on lang forum:)  If what I tell unsoud (there could be 2 possible decisions for a compiler which cannot be decided), then anything else is useless. Like uses cases and opinions.

My use case is the ability to work with result-based error code and exception raising uniformly, mapping one into another with ease whatever is preferred.

**Reply by CyrusNajmabadi** [reply]
> What you could do is prove that the inheritance rule is unsound and will not work. That kind of talk I expect on lang forum:)

Language design is far more than just deciding if the rules are sound it not. We have to consider if they're actually sensible. 

> So rule can be

I genuinely don't know why that rule would be preferable. You haven't given a use case to demonstrate why it's better, just that you want it for your particular domain (which still doesn't have any real examples present). I get this provides *a* solution for your A/B/C/D case, but we haven't even gotten a reason that that case is interesting it important, let alone an intuition around it that would indicate that this solution would be the preferred one.

**Reply by CyrusNajmabadi** [reply]
> My use case is the ability to work with result-based error code and exception raising uniformly, mapping one into another with ease whatever is preferred.

Can you show the actual API you want here?

It's unclear to be still why methods to map between domains is not sufficient.

**Reply by CyrusNajmabadi** [reply]
> But because there is one and only one possible inheritance

This is not true. The language supports multiple inheritance for some types for example.

**Reply by dzmitry-lahoda** [reply]
hm, just googled. it tells nor csharp nor cts nor clr have support for multiple inheritance, to break the rule. there is one base type in api. also not aware of supported proposals for that.  cpp types are flatouted into structs.

**Reply by dzmitry-lahoda** [reply]
> It's unclear to be still why methods to map between domains is not sufficient

so you know cases in which can have Cast/Into/To/Convert/Map like methods to get instance of one type from other? 

but if there is context of expected return type and implicits, why would not to avoid calling methods if type can be decided? less typing and noise in code.

actually type is decided when there is one implicit in child only.  but as soon as parent introduces implicit is will break child compilation. adding implicits is in theory is backward incomatible, so one may protect by doing sealed or internal until full api is ready.

**Reply by CyrusNajmabadi** [reply]
> so you know cases in which can have Cast/Into/To/Convert/Map like methods to get instance of one type from other?

Presumably the places you were adding these conversion operators. I would just make them named methods

**Reply by CyrusNajmabadi** [reply]
> hm, just googled. it tells nor csharp nor cts nor clr have support for multiple inheritance,

C# absolutely does. For example, interfaces have MI.

**Reply by CyrusNajmabadi** [reply]
> why would not to avoid calling methods if type can be decided? less typing and noise in code.

I think a core difference is that you see this as noise, whereas I see it as clarity. I'm ok removing things if it makes things clearer. However, in this case, it genuinely seems to obfuscate. Indeed, I'm not even sure what these conversions would do. Having multiple ways to get to the same end point that literally may result in different values is exactly the sort of thing that shouldn't happen automatically as it just means surprises when you ends up with one and not the other. Being explicit in code means you know exactly what path was taken and there is no guessing.

**Reply by dzmitry-lahoda** [reply]
3. there should be guidance telling to that implicit operators should be coded in predictable way. actually, in case provided they do predictable way. and actually there is already guidance to make implicit predictable.

**Reply by CyrusNajmabadi** [reply]
Can you show how you would implement this to make the separate impls work in a consistent fashion?

**Reply by dzmitry-lahoda** [reply]
> I think a core difference is that you see this as noise, whereas I see it as clarity. I'm ok removing things if it makes things clearer. However, in this case, it genuinely seems to obfuscate. Indeed, I'm not even sure what these conversions would do. Having multiple ways to get to the same end point that literally may result in different values is exactly the sort of thing that shouldn't happen automatically as it just means surprises when you end up with one and not the other. Being explicit in code means you know exactly what path was taken and there is no guessing.

Most of C# innovation is removing noise. There are almost zero other innovations after generics were introduced in CLR 2.0. Everything is decompiled to basics and could be done manually with a huge boilerplate in old C#. 

Lack of expressiveness in C# leads to ASP, WPF, WCF, etc. Where everything is `object` and many ways to do the same things.

C# cannot be explicit, as it sugar on top of sugar and huge pile of reflection on top.

So these generic statements are opinions contradicting what C# is and .NET ecosystem is(there are many ways to do the same thing, you never know what is going on, everything is magic and guessing). With a lack of precise definitions of things and ignorance of what is accepted by `multiple inheritances`.

I cannot argue against these.

**Reply by dzmitry-lahoda** [reply]
In theory, implementation is possible which works when there are multiple true inheritances or not. And it is solvable by forming a graph of all pathways from T1 to TN and following the rule of doing inheritance first or panic in case of no inheritance with impilcits ambiguity.  Correct implementation of a pattern is guided to make sure people do not implciit to wrong things.

Same guidance as it is done for disposal, finalizers, visibility, constructors, single inheritance implicit, etc.

I do not see why to implement anything until there is consensus that that would be accepted as a patch. Probably we should wait for more years until the .NET crowd grows up.

**Reply by CyrusNajmabadi** [reply]
> Most of C# innovation is removing noise. 

That's not what i said though.  I said that removing this removes *clarity*.  Generally, we don't remove things if the absence of them would lead them to be more confusing.

> There are almost zero other innovations after generics were introduced in CLR 2.0. Everything is decompiled to basics and could be done manually with a huge boilerplate in old C#.

That doesn't change anything here.  That we add "sugar" features when we think it's a *good idea* doesn't mean that every feature gets done if it is just sugar.  The idea still needs to be a good one, and that means addressing concerns about things like clarity.

> So these generic statements are opinions contradicting what C# is and .NET ecosystem is(there are many ways to do the same thing, you never know what is going on, everything is magic and guessing). 

There is no contradiction here.  C# allows the things it allows because we (the language designers) assess the overall value we think the features provide, weighting their positives against their negatives.  If we think there is enough value, and the costs are in line with that, then we go ahead with it.  If we think that the negatives outweigh the positives, or that the costs don't justify the benefits, then we don't go ahead with it.  Arguing that we've done features in the past, and thus we should do this feature isn't going to go anywhere.  We need to see that making *this* particular change would be a good thing, and that means addressing core concerns, like the ones i've brought up above.

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda 

> I do not see why to implement anything

I was asking you to show how you would implement your operators such taht this constraint was held:

"there should be guidance telling to that implicit operators should be coded in predictable way."

I'm asking for a fleshed out example showing how you would do that.  The reason i'm asking for this is that the example you gave is simply so abstract that it can't be usefully looked at as a motivating example to help drive any sort of intuitions here on what sort of changes would be right/wrong for the language.  So, asking to flesh these examples out is to take it from just being an abstract talk of 'A/B/C/D' and to make it into a much more grounded discussion where we can look at how this would play out in practice with a realistic API to ensure that the changes being discussed would be appropriate.

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda 

> Probably we should wait for more years until the .NET crowd grows up.

This behavior goes against our Code of Conduct.  Please be respectful and act in a constructive and professional manner.   Telling team members to "grow up" because you don't like their assessment, is not ok.  Thanks!

**Reply by CyrusNajmabadi** [reply]
> In theory, implementation is possible which works when there are multiple true inheritances or not. And it is solvable by forming a graph of all pathways from T1 to TN and following the rule of doing inheritance first or panic in case of no inheritance with impilcits ambiguity. 

Again, you are stating a solution here.  But the point i'm making is that i don't see anything that justifies why *this* solution is actually a good idea.  As i've pointed out myself, it's not at all intuitive why this would be the desired path, and i'm not seeing how this would play out in practice because you haven't demonstrated what this bit would look like:

> Correct implementation of a pattern is guided to make sure people do not implciit to wrong things.

Can you please show a realistic example here, that also shows what that correct implementation would look like?  You claim it exists, but it's not readily apparent to me what that would look like, so we def need that fleshed out to continue the discussion.  Thank you.

**Reply by dzmitry-lahoda** [reply]
Sure, thanks. You may close question if that possible

**Reply by dzmitry-lahoda** [reply]
> This behavior goes against our Code of Conduct. Please be respectful and act in a constructive and professional manner. Telling team members to "grow up" because you don't like their assessment, is not ok. Thanks

Not sure what you mean. Just let I clarify. Grows means when people start liking types, not reflection. Like Java grown into Kotlin. .NET is 10 years younger.

**Reply by dzmitry-lahoda** [reply]
Hmm, I do not see a button to close the question. So will link it in our code base as the reason why we do not have nice sugar to simplify things:)

**Reply by CyrusNajmabadi** [reply]
> Just let I clarify. Grows means when people start liking types, 

.net and c# already embrace static typing heavily. That is unrelated to whether or not we think these sorts of rules should be in effect. Rules need to add more value than potential problems.

**Reply by CyrusNajmabadi** [reply]
> Hmm, I do not see a button to close the question.

There is no need to close the discussion. It is still waiting on you for answers to the questions raised above. 

For example:

> Can you please show a realistic example here, that also shows what that correct implementation would look like? You claim it exists, but it's not readily apparent to me what that would look like, so we def need that fleshed out to continue the discussion. Thank you.

**Reply by dzmitry-lahoda** [reply]
example, 
1. you have one class hierarchy formed by some framework. so you want your own hierarchy. there is always morphism from your hierarchy to framework one. 
2. you have 2 representations of each object. one is good in one bounded context and the other in one, there is mutual morphism amid. Your source generates both hierarchies having the same data, so different representations.

**Reply by dzmitry-lahoda** [reply]
i have unsubscribed from thread

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda I'm asking you to show the actual code for how you would write these implicit operators and have them follow the desired semantics you showed above. Specifically how you ensure that any of the N paths that are possible result in the same value and type and runtime.

**Reply by CyrusNajmabadi** [reply]
> Your source generates 

If you have a source generator, why can you not just generate the conversion methods between these two hierarchies?

**Reply by dzmitry-lahoda** [reply]
Prove can exist only about that type path is unique (and reasonable within single inheritance realm of csharp/clr). Prove does not prove that that path is not garbage. Garbage in garbage out in any language.  Csharp has no type system to help prove, only guidances. Code with guidance is in the head of the thread (actually that is the essence of generated code).

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda sorry, I'm not sure what your trying to say.

**Reply by dzmitry-lahoda** [reply]
> Specifically how you ensure that any of the N paths that are possible result in the same value and type and runtime.

This is not needed and is impossible. Never asked for that.

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda you said:

> Correct implementation of a pattern is guided to make sure people do not implciit to wrong things.

I'm asking what that guidance would be and what the correct implementation of this pattern would be in this example you're presenting here.  Basically, it's not readily apparent how that would work, so i'm basically stating: "please show your work" so that i can better understand what you're proposing.

Thanks!

**Reply by CyrusNajmabadi** [reply]
@dzmitry-lahoda 

As before, can you please show a realistic example of these hierarchies, and how your implicit operators would work here?  Having vague A/B/C/D cases isn't helpful as it's too divorced from real work cases.  We need something concrete to look at to get an intuition on how this would work and why the proposed rules and guidance would be suitable and appropriate.

> This is not needed and is impossible. Never asked for that.

@dzmitry-lahoda You asked for changes to the language.  Part of the process around that is showing real examples of how this would work out so we can assess the value here and understand the pros and cons.  Asking for examples is not impossible, nor is it unreasonable to ask for this given that you said:

> there should be guidance telling to that implicit operators should be coded in predictable way. actually, in case provided they do predictable way. and actually there is already guidance to make implicit predictable.

I'm asking specifically what that guidance would be and how it would be applied in the context of a real world example.  Right now your example is too vague, and every attempt at understanding this space better has been met with resistance and a seeming unwillingness to on your part to provide any more information.  If you would like to see any change happen here, please provide this information as otherwise this just isn't actionable in its current form.

**Reply by dzmitry-lahoda** [reply]
I am ok with the answers. I not going to make a proposal as it requires all that you said. Examples, rules, and guidance. In this case, in can be a waste of effort as it is not yet known if others could hit same limit and will try same design and will desire same solution. So better just wait for votes, at lost for me.
